vector-stores/spring-ai-mongodb-store/src/main/java/org/springframework/ai/vectorstore/MongoDBVectorStore.java

```diff
@@ -0,0 +1,128 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.ai.vectorstore;
+
+import com.mongodb.BasicDBObject;
+import org.springframework.ai.document.Document;
+import org.springframework.ai.embedding.EmbeddingClient;
+import org.springframework.ai.vectorstore.impl.InMemoryVectorStore;
+import org.springframework.data.mongodb.core.MongoTemplate;
+import org.springframework.data.mongodb.core.query.Query;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static org.springframework.data.mongodb.core.query.Criteria.where;
+
+/**
+ * @author Chris Smith
+ */
+public class MongoDBVectorStore implements VectorStore {
+
+	MongoTemplate mongoTemplate;
+
+	EmbeddingClient embeddingClient;
+
+	private final String VECTOR_COLLECTION_NAME = "vector_store";
+
+	public MongoDBVectorStore(MongoTemplate mongoTemplate, EmbeddingClient embeddingClient) {
+		this.mongoTemplate = mongoTemplate;
+		this.embeddingClient = embeddingClient;
+		if (!mongoTemplate.collectionExists(VECTOR_COLLECTION_NAME)) {
+			mongoTemplate.createCollection(VECTOR_COLLECTION_NAME);
+		}
+	}
+
+	/**
+	 * Maps a basicDBObject to a Spring AI Document
+	 * @param basicDBObject
+	 * @return
+	 */
+	public Document mapBasicDbObject(BasicDBObject basicDBObject) {
+		String id = basicDBObject.getString("_id");
+		String content = basicDBObject.getString("text");
+		Map<String, Object> metadata = (Map<String, Object>) basicDBObject.get("metadata");
+		List<Double> embedding = (List<Double>) basicDBObject.get("embedding");
+
+		Document document = new Document(id, content, metadata);
+		document.setEmbedding(embedding);
+
+		return document;
+	}
+
+	@Override
+	public void add(List<Document> documents) {
+		for (Document document : documents) {
+			List<Double> embedding = this.embeddingClient.embed(document);
+			document.setEmbedding(embedding);
+
+			mongoTemplate.save(document, VECTOR_COLLECTION_NAME);
+		}
+	}
+
+	@Override
+	public Optional<Boolean> delete(List<String> idList) {
+		Query query = new Query(where("_id").in(idList));
+
+		var deleteRes = mongoTemplate.remove(query, VECTOR_COLLECTION_NAME);
+		long deleteCount = deleteRes.getDeletedCount();
+
+		return Optional.of(deleteCount == idList.size());
+	}
+
+	@Override
+	public List<Document> similaritySearch(String query) {
+		return this.similaritySearch(query, 4);
+	}
+
+	@Override
+	public List<Document> similaritySearch(String query, int k) {
+		List<Double> queryEmbedding = this.embeddingClient.embed(query);
+
+		// TODO: explore making this query better
+		return this.mongoTemplate.findAll(BasicDBObject.class, VECTOR_COLLECTION_NAME)
+			.stream()
+			.map(this::mapBasicDbObject)
+			.map(entry -> new InMemoryVectorStore.Similarity(entry.getId(),
+					InMemoryVectorStore.EmbeddingMath.cosineSimilarity(queryEmbedding, entry.getEmbedding())))
```

Comment:
I'm confused. I've assumed, wrongly perhaps, that MongoDB provides a build-in Vector Search support? 

If it doesn't provide a native vector search capability, then I'm not convinced this Mongo DB sore would perform or scale any better than the in-memory one.  
@markpollack what do you think?

Reply:
Thanks for taking the time to review @tzolov!

The managed version of MongoDB (atlas), allows configuration to provide semantic search capability by automatically attaching embeddings to documents using  fields configured for each collection.
https://www.mongodb.com/products/platform/atlas-vector-search

Given that, I more of less went for making a `PersistentVectorStore` backed by MongoDB, partially out of "could I do it" and to provide another persistent  (though not optimal) out of the box.